src/renderers/dom/client/MarkupMismatchError.js

```diff
@@ -0,0 +1,216 @@
+ * present in the server markup.
+ * @param {DOMElement} the DOM element from the server markup that is missing an attribute
+ * @param {String} the attribute name
+ * @param {String} the attribute value in the client component tree.
+ */
+function throwAttributeAddedMismatchError(node, attr, value) {
+  throw new MarkupMismatchError(`The attribute '${attr}' was added.`, node, `<No ${attr} value>`, `${attr}=${value}`);
+}
+
+/**
+ * throw an error when there is a DOM attribute in the client component tree that has a
+ * different value in the server markup.
+ * @param {DOMElement} the DOM element from the server markup that has the changed attribute
+ * @param {String} the attribute name
+ * @param {String} the attribute value in the server markup
+ * @param {String} the attribute value in the client component tree.
+ */
+function throwAttributeChangedMismatchError(node, attr, serverValue, clientValue) {
+  throw new MarkupMismatchError(
+    `The value of the attribute '${attr}' differed.`,
+    node,
+    `${attr}=${serverValue}`,
+    `${attr}=${clientValue}`);
+}
+
+/**
+ * throw an error when html set with dangerouslySetInnerHTML has a
+ * different value in the server markup and client component tree.
+ * @param {DOMElement} the DOM element from the server markup that has the dangerouslySetInnerHTML
+ * @param {String} the dangerouslySetInnerHTML value in the server markup
+ * @param {String} the dangerouslySetInnerHTML value in the client component tree.
+ */
+function throwInnerHtmlMismatchError(node, serverValue, clientValue) {
+  throw new MarkupMismatchError(
+    'The value for dangerouslySetInnerHTML differed.',
+    node,
+    serverValue,
+    clientValue
+  );
+}
+
+/**
+ * throw an error when the text content of a node has a
+ * different value in the server markup and client component tree.
+ * @param {DOMElement} the DOM element from the server markup that has differing text
+ * @param {String} the text value in the server markup
+ * @param {String} the text value in the client component tree.
+ */
+function throwTextMismatchError(node, serverValue, clientValue) {
+  throw new MarkupMismatchError('Text content differed.', node, `'${serverValue}'`, `'${clientValue}'`);
+}
+
+/**
+ * throw an error when the component type (i.e. element, text, empty) of a node has a
+ * different value in the server markup and client component tree.
+ * @param {DOMNode|Array<DOMNode>} the DOM element or elements from the server
+ *   markup that is/are different on client
+ * @param {String} a user-readable description of the component in the client
+ *   component tree.
+ */
+function throwComponentTypeMismatchError(node, clientComponentDesc) {
+  var serverValue = '';
+
+  if (Array.isArray(node) && node.length === 3) {
+    serverValue = `A text node with text: '${abridgeContent(node[1].textContent)}'`;
+  } else if (Array.isArray(node) && node.length === 2) {
```

Comment:
Sorry, why does a node array of length 2 imply that it is a blank text node?  Why does it being an array imply that it is a text node?  I feel like this logic conflicts with the jsdocs/description above.